packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,211 @@
+    group,
+    addressIndex,
+    skipAddressIndexes = []
+  }: GenerateAddressProps): SensitiveAddressData => {
+    if (group !== undefined && (!Number.isInteger(group) || group < 0 || group >= TOTAL_NUMBER_OF_GROUPS))
+      throw new Error(`Keyring: Could not generate address in group ${group}, group is invalid`)
+
+    if (addressIndex !== undefined) {
+      if (!Number.isInteger(addressIndex) || addressIndex < 0)
+        throw new Error(`Keyring: Could not generate address, ${addressIndex} is not a valid addressIndex`)
+
+      if (group !== undefined || skipAddressIndexes.length > 0)
+        throw new Error(
+          'Keyring: Could not generate address, invalid arguments passed: when addressIndex is provided the group and skipAddressIndexes should not be provided.'
+        )
+
+      return this._deriveAddressAndKeys(addressIndex)
+    }
+
+    const initialAddressIndex = 0
+
+    let nextAddressIndex = skipAddressIndexes.includes(initialAddressIndex)
+      ? findNextAvailableAddressIndex(initialAddressIndex, skipAddressIndexes)
+      : initialAddressIndex
+    let newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+
+    while (group !== undefined && groupOfAddress(newAddressData.hash) !== group) {
+      nextAddressIndex = findNextAvailableAddressIndex(newAddressData.index, skipAddressIndexes)
+      newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+    }
+
+    return newAddressData
+  }
+
+  private _getNonSensitiveAddressData = ({
+    hash,
+    index,
+    publicKey
+  }: SensitiveAddressData): NonSensitiveAddressData => ({ hash, index, publicKey })
+
+  private _initFromMnemonic = (mnemonic: Uint8Array | null, passphrase: string) => {
+    if (this.root) throw new Error('Keyring: Secret recovery phrase already provided')
+    if (!mnemonic) throw new Error('Keyring: Secret recovery phrase not provided')
+
+    const seed = metamaskBip39.mnemonicToSeedSync(mnemonic, wordlist, passphrase)
+    this.hdWallet = HDKey.fromMasterSeed(seed)
+    this.root = this.hdWallet.derive(this.hdPath)
+
+    passphrase = ''
+  }
+
+  private _deriveAddressAndKeys = (addressIndex: number): SensitiveAddressData => {
+    if (!this.root) throw new Error('Keyring: Cannot derive address and keys, secret recovery phrase is not provided')
+    if (!isAddressIndexValid(addressIndex)) throw new Error('Invalid address index path level')
+
+    const keyPair = this.root.deriveChild(addressIndex)
+
+    if (!keyPair.publicKey) throw new Error('Keyring: Missing public key')
+    if (!keyPair.privateKey) throw new Error('Keyring: Missing private key')
+
+    const publicKey = bytesToHex(keyPair.publicKey)
+    const privateKey = keyPair.privateKey
+    const hash = blake.blake2b(Uint8Array.from(keyPair.publicKey), undefined, 32)
+    const type = new Uint8Array([0])
+    const bytes = new Uint8Array(type.length + hash.length)
+    bytes.set(type, 0)
+    bytes.set(hash, type.length)
+    const address = bs58.encode(bytes)
```

Comment:
We used to generate this info using `Buffer`, now we use `Uint8Array`: https://github.com/alephium/alephium-frontend/blob/master/packages/shared-crypto/lib/wallet.ts#L111C1-L118C1

```ts
  const publicKey = keyPair.publicKey.toString('hex')
  const privateKey = keyPair.privateKey.toString('hex')
  const hash = blake.blake2b(Uint8Array.from(keyPair.publicKey), undefined, 32)
  const pkhash = Buffer.from(hash)
  const type = Buffer.from([0])
  const bytes = Buffer.concat([type, pkhash])
  const address = bs58.encode(bytes)
```